### ISSUE TITLE
Clean up the Tensor class.

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#16465]
The `Tensor` class has a somewhat complicated structure that is due to the fact that we build it recursively: a `Tensor<rank,dim>` internally consists of an array of `Tensor<rank-1,dim>` objects. It might have been natural to build this on a `Tensor<1,dim>` as the base case for the recursion (and I think that's how it was at some point) but in reality, it is built on `Tensor<0,dim>` as the base case. This introduces a number of complexities because we need to use dispatch to overloaded functions in some places. This is tough to read, and probably also slow to compile. But we can do better today with `if constexpr` expressions.

If we do this right, we can also do a number of other simplifications:
* Instead of storing an array of `Tensor<0,dim>` in `Tensor<1,dim>`, I would like to just store an array of numbers.
* Instead of storing a `Number[dim>0 ? dim : 1]` array (to ensure non-zero size), I would like to make this a `std::array<Number,dim>`.
* Once we're there, my final step is to convert the underlying array for `Tensor<1,dim>` to a `VectorizedArray` object. This would have the effect that for things such as `operator+`, `operator*(double)`, etc, we can use vectorized instructions. I think this will be more efficient than the loops over elements we currently have.

**Comment by tamiko** [comment]
@bangerth Yes, we had `Tensor<1,dim>` as the base for the recursion at one point. But then we had an entirely separate `Tensor<0, dim>` specialization duplication everything again. I changed this to the current recursive pattern with `Tensor<0,dim>` back in 2015 (?). Unfortunately, most of the awkwardness of the interface had to be maintained at that point.

**Comment by bangerth** [comment]
@tamiko No worries -- we didn't have the language tools at the time to deal with this.

**Comment by bangerth** [comment]
So this is awkward, and I might need help from our vectorization specialists (@kronbichler @peterrum ?)

I am trying something like this for the elements of a rank-1 tensor:
```
  /**
   * Whether or not the 'values' array is stored as a vectorized array
   * (of possibly more than 'dim' elements, which would then have to
   * be ignore by all functions).
   */
  static constexpr bool store_values_as_vectorized_array =
    ((rank_ == 1) && (dim <= 4) &&
     (std::is_same_v<Number, double> || std::is_same_v<Number, float>)&&(
       internal::VectorizedArrayWidthSpecifier<Number>::max_width >= 4));

  /**
   * Array of tensors holding the elements of the tensor. If this is
   * a rank-1 tensor, then we simply need an array of scalars.
   * Otherwise, it is an array of tensors one rank lower.
   *
   * If we can use vectorization, then the array may actually be holding more
   * than `dim` elements. In that case, the remainder of the elements are
   * padded by zeros, and all operations must make sure that this zero
   * padding remains in place to ensure that we can use efficient vector
   * operations.
   */
  std::conditional_t<rank_ == 1,
                     std::conditional_t<store_values_as_vectorized_array,
                                        VectorizedArray<Number, 4>,
                                        std::array<Number, dim>>,
                     std::array<Tensor<rank_ - 1, dim, Number>, dim>>
    values;
```
This seems reasonable, but we run into trouble because the constructor of `Tensor` is marked as `constexpr` and so can only call other `constexpr` classes. Unfortunately, pretty much nothing in `vectorization.h` is `constexpr`. So I added a couple `constexpr` annotations to the corresponding constructors of `VectorizerArray<float/double, 4>`, but this leads to the following error:
```
13647: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/base/vectorization.h: In constructor ‘constexpr dealii::VectorizedArray<float, 4>::VectorizedArray(float)’:
13647: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/base/vectorization.h:2042:3: error: member ‘dealii::VectorizedArray<float, 4>::data’ must be initialized by mem-initializer in ‘constexpr’ constructor
13647:  2042 |   }
13647:       |   ^
13647: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/base/vectorization.h:2250:10: note: declared here
13647:  2250 |   __m128 data;
13647:       |          ^~~~
13647: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/base/vectorization.h: In member function ‘constexpr dealii::VectorizedArray<float, 4>& dealii::VectorizedArray<float, 4>::operator=(float) &’:
13647: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/base/vectorization.h:2059:23: error: call to non-‘constexpr’ function ‘__m128 _mm_set1_ps(float)’
13647:  2059 |     data = _mm_set1_ps(x);
13647:       |            ~~~~~~~~~~~^~~
```

The first of these we *could* address by initializing the `__m128` data object, though we probably don't want to. The second one, however, is a real issue. 

What should we do here?

**Comment by kronbichler** [comment]
This appears to be a tricky problem. I agree that we should not initialize the `__m128` data object because many places depend upon it in terms of performance. I was wondering how `std::array` is doing it in terms of not initializing items, but I couldn't see it. I'm wondering if this might be the time where we try to rework `VectorizedArray` into a more generic array class that only uses casts to the intrinsics types `__m128, __m256, __m512, __m128d` etc for implementing the operation but not for storage, as discussed with @tamiko a few years ago, see #11719. I am not entirely sure if it catches all cases with the same performance, but we might give it a try.

Regarding the second case, I need to think about it. Is there a way we can check whether we are in the `constexpr` path (and where we would select a loop over the array elements) and the old path (where we can use `_mm_set1_ps`)?

**Comment by bangerth** [comment]
Unlike `consteval`, there is no way whether the compiler is current evaluating a `constexpr` exprssion :-(

For the constructor, though, one possible way to do that would be to have another constructor that takes a dummy argument to indicate that it is initializing the values. A bit like `std::in_place` is used to indicate something in constructor (6) here: https://en.cppreference.com/w/cpp/utility/optional/optional .

A problem I foresee is that many of the functions that would benefit substantially from vectorization, say
```
template <int rank_, int dim, typename Number>
template <typename OtherNumber>
constexpr inline DEAL_II_ALWAYS_INLINE
  DEAL_II_HOST_DEVICE Tensor<rank_, dim, Number>                     &
  Tensor<rank_, dim, Number>::operator+=(
    const Tensor<rank_, dim, OtherNumber> &p)
{
  for (unsigned int i = 0; i < dim; ++i)
    values[i] += p.values[i];
  return *this;
}
```
are marked as `constexpr`. But because (apparently) the compiler intrinsics are all not `constexpr`, we can't call them. That's a real bummer. Not sure what to do about that.

**Comment by bangerth** [comment]
I misremembered the `consteval` thing. You *can* determine whether something is `constexpr` evaluated, but only since C++20: https://stackoverflow.com/questions/46919582/is-it-possible-to-test-if-a-constexpr-function-is-evaluated-at-compile-time

**Comment by masterleinad** [comment]
> Instead of storing a Number[dim>0 ? dim : 1] array (to ensure non-zero size), I would like to make this a std::array<Number,dim>.

This will likely not work on GPUs.

**Comment by bangerth** [comment]
Hm, but this change already happened in #16474 and it compiled just fine...

**Comment by masterleinad** [comment]
> Hmm... I might try manually for NVIDIA, AMD and Intel GPUs.

See https://godbolt.org/z/YcePz7T9P for example (basically we need `-expt-relaxed-constexpr` for this to work).

**Comment by bangerth** [comment]
Are we actually running into this with our current code base? #16474 did pass all of our tests, and as far as I know none of the periodic testers have completely broken either.

**Comment by masterleinad** [comment]
Our CI enforces that flag since it avoids a lot of annoying compiler warnings but its not used by default in `Kokkos` or `nvcc`. It seems to be fine with `clang++` for `Cuda` and with `SYCL` for Intel GPUs and since `hipcc` is based on top of `clang++`, I expect it to also be fine on AMD GPUs.
Thus, we probably just need to enforce that `Kokkos` was configured with `Kokkos_ENABLE_CUDA_CONSTEXPR=ON`.

**Comment by masterleinad** [comment]
> I don't have any experience with GPUs or any of these compilers. Do we need to roll back all of these patches, or do you think you can finagle things in such a way that we can make sure it works?

See #16531.

**Comment by bangerth** [comment]
I've given this some more thought. What I really want to do is this:
```
  /**
   * Whether or not the 'values' array is stored as a vectorized array
   * (of possibly more than 'dim' elements, which would then have to
   * be ignored by all functions), or simply as a std::array.
   *
   * We can store the values as a vectorized array if this is a rank-1
   * tensor, and if it stores `double` or `float` values, and if there are
   * at most 4 values -- we do not try to vectorize the uncommon case of
   * `dim>4`. If we decide to store values this way, then this also
   * propagates to the sub-tensors that make up the sub-tensors of higher
   * rank tensors.
   */
  static constexpr bool store_value_as_vectorized_array =
    ((rank_ == 1) && (dim <= 4) &&
     (std::is_same_v<Number, double> || std::is_same_v<Number, float>)&&(
       internal::VectorizedArrayWidthSpecifier<Number>::max_width >= 4));

  /**
   * The array holding the subelements. If this is a rank-1 tensor, then
   * the array simply needs to store some scalars. If it is a higher-rank
   * tensor, then we need an array of sub-tensors.
   */
  std::conditional_t<rank_ == 1,
                     std::conditional_t<store_value_as_vectorized_array,
                                        VectorizedArray<Number, 4>,
                                        std::array<Number, dim>>,
                     std::array<Tensor<rank_ - 1, dim, Number>, dim>>
    values;
```
In other words, I'd like to store the elements of a rank-1 tensor as a `VectorizedArray`. Doing so would allow replacing loops over the `dim` elements of tensors by calls to intrinsics that are, presumably, more efficient. Given how much tensor arithmetic we do, this would suggest to be a substantial win.

The problem we run into is basically this:
* Plenty of functions of `Tensor<rank,dim>` are `constexpr`, including the constructor, `operator[]`, and a whole bunch of other functions and operators. But if we store the elements of the tensor as the elements of a `VectorizedArray`, at the least we need that the constructor and `operator[]` of that latter class are also `constexpr`.
* Alas, they are not. In fact, they can't be: `operator[]` contains `reinterpret_cast<double*>(&data)` to access the elements of the `__m256d` (or similar) data type because there is no other way to access the elements of such types as far as I know. Most of the other functions call intrinsics such as `_mm256_add_pd` that aren't `constexpr` either. (Some of the `VectorizedArray` member functions use direct operators when `DEAL_II_COMPILER_USE_VECTOR_ARITHMETICS` is set, but it's not clear to me that these are `constexpr`.) I don't see how we can make `VectorizedArray` be sufficiently `constexpr`-safe.
* So at least with the current design, `Tensor` cannot build on `VectorizedArray`.
* I did look into `std::simd` as a replacement. That's not a standard class, but we could see if we can use it if a compiler provides it. Alas, again, it turns out that the specification of `std::simd` does not include `constexpr` specifications: https://en.cppreference.com/w/cpp/experimental/simd/simd Since we have even less control over `std::simd` than over `VectorizedArray`, that seems like a dead-end too.
* I did also look into `std::valarray` as a replacement. That class has been around for a long time, and I think that at the time of its invention, people thought of it as something that compilers would vectorize or otherwise optimize. But `std::valarray` also does not have `constexpr` specifications.

So this somehow ends my quest for this specific optimization, I fear. The only other option I see is remove the `constexpr` specifications from `Tensor` again. Perhaps that worth it -- what do you think?

**Comment by kronbichler** [comment]
Thank you for the detailed list. I have another idea: What if we use the following declaration of `Tensor`:
```c++
class alignas(/* some constexpr that is either 1, 2 or 4*/ * sizeof(Number)) Tensor
```
then set
```c++
  std::conditional_t<rank_ == 1,
                     std::conditional_t<store_value_as_vectorized_array,
                                        std::array<Number, /*either 1, 2, 4 or dim*/>>,
                     std::array<Tensor<rank_ - 1, dim, Number>, dim>>
    values;
```
and then in the arithmetic operations cast `&values[0]` to `VectorizedArray<Number>*` and use the vectorized-array arithmetic operations in `operator+`, `operator*` etc? The cast is possible because we have designed the class to be of POD type (more precisely, I think we need it to be a trivial type).

What do you think? I could implement the part that makes the casts (and check the generated code), if you would be wanting to make array sizes compatible with powers of 2.

**Comment by bangerth** [comment]
Actually, I had thought about that and decided that it was too obscure an approach to suggest :-) But I had tried this and ran into a number of road blocks that I also don't know how to circumvent. First, the cast to `VectorizedArray<Number>*` is a `reinterpret_cast`, which is not allowed in `constexpr` functions :-( (And if you ask whether using a `union` is allowed -- no, not either in the way we would have to here.) Second, even if it worked, there still is the issue that the compiler intrinsics are all not `constexpr` either.

In other words, I do not actually think that any of the ideas we have been able to come up with are compatible with `constexpr`. We will have to think what it is we want more: Compile-time evaluation, or the ability to use vector intrinsics.

**Comment by masterleinad** [comment]
There is at least [bit_cast](https://en.cppreference.com/w/cpp/numeric/bit_cast) in `C++20` that allows `constexpr` reinterpretations of memory.

**Comment by bangerth** [comment]
`bit_cast` creates a *copy* of the object, whereas `reinterpret_cast` simply reinterprets a memory location :-( Either way, the problem with the non-`constexpr`ness of the intrinsics remains.

**Comment by bangerth** [comment]
We've done here whatever could be done. #16848 and #16865 showed that trying to manually use vector intrinsics does not result in code that's better than what compilers can do automatically.
